Grader: Add more checks to `fork-wait-exit`

Blazefrost opened the pull request:
This commit introduces two more checks to the `fork-wait-exit` assignment.

* The first check asserts whether the `wait` implementation handles unmapped virtual addresses and maps pages accordingly. If the syscall does not handle it properly, Selfie will probably raise a SEGFAULT signal.
* The second check asserts whether `wait` handles invalid addresses outside the 4GiB correctly and returns -1 instead of a terminated child's PID.

Blazefrost commented:
There is a check for illegal addresses still pending if there is enough time left before the assignment is published. 

Also, when an invalid pointer is passed to `wait`, a zombie's status should not be consumed, right? I could extend the second check to assert this behavior.

@fischer-martin could you check it out?

ckirsch commented:
> There is a check for illegal addresses still pending if there is enough time left before the assignment is published.
> 
> Also, when an invalid pointer is passed to `wait`, a zombie's status should not be consumed, right? I could extend the second check to assert this behavior.
> 
> @fischer-martin could you check it out?

Sounds all good to me. @Blazefrost Is the PR ready or are you still working on it?

Blazefrost commented:
I would like to extend the invalid pointer check as aforementioned and maybe add a check whether `wait` can handle NULL pointer and just return the PID in this case, if that's okay with you.

ckirsch commented:
> I would like to extend the invalid pointer check as aforementioned and maybe add a check whether `wait` can handle NULL pointer and just return the PID in this case, if that's okay with you.

Sure! Please go ahead.

Blazefrost commented:
Done! However, I probably need to check whether an implementation maps a page to NULL. I thought about implementing it by trying to access NULL and provoking a segfault.

@fischer-martin what do you think about the current state of the PR?

fischer-martin commented on grader/assignments/fork-wait/invalid-address.c:
```diff
@@ -0,0 +1,44 @@
+  uint64_t pid;
+  uint64_t* code;
+
+  code = malloc(8);
```
The variable name is misleading. Something like `heap_top` would probably fit better.

fischer-martin commented on grader/assignments/fork-wait/null-ptr.c:
```diff
@@ -0,0 +1,15 @@
+  pid = fork();
+  if (pid == 0)
+    exit(0);
+  else {
```
The curlies can be removed.

fischer-martin commented on grader/assignments/fork-wait/unmapped-page-wait.c:
```diff
@@ -0,0 +1,28 @@
+  code = malloc(8 * 4096);
+
+  pid = fork();
+  if (pid == 0) {
```
The curlies can be removed (same for `else`).

fischer-martin commented:
@Blazefrost Looks really good! I've only added some annotations regarding cosmetic changes.

ckirsch commented:
@Blazefrost The travis build still fails though in the grader target...

Blazefrost commented:
I fixed the redundant braces and fixed the unit test failure. The failure was caused by a check whether the expected exit code overlaps with one of Selfie's exit codes. I also adapted the checks so that they do no longer assume the child PID is necessarily 2.

The additional checks should be done now!
